### ISSUE TITLE
Updating an object in a list with updateOne makes it an array

### ISSUE THREAD

**Issue by kimbylr** [Automattic/mongoose#14000]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

7.6.2

### Node.js version

18

### MongoDB server version

6

### Description

I'm updating an object (`item`) in an array (`items`) using `$set` (cf. code below).

After upgrading to the latest version of Mongoose, the updated object gets "wrapped" in an array, i.e. `[{... data ...}]` instead of `{... data ...}`.

![Bildschirmfoto 2023-10-23 um 21 20 35](https://github.com/Automattic/mongoose/assets/21242860/124de1ab-a58c-4893-a3d4-da90eddfcad6)

Mongoose 7.6.1 is the last version where this does not happen.

Glad to help if more info is needed.

### Steps to Reproduce

Sample code:

```
await List.updateOne(
  { _id: listId, 'items._id': item._id },
  { $set: { 'items.$': newItem } },
);
```

### Expected Behavior

The object with id `item._id` in array `items` is replaced with `newItem`.

**Comment by IslandRhythms** [comment]
Please modify the script below to demonstrate your issue

```javascript
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema({
  items: [{

  }]
});
const id = new mongoose.Types.ObjectId();
const Test = mongoose.model('Test', testSchema);

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  const entry = await Test.create({
    items: {
      _id: id,
      first: 'Test',
      last: 'Testeron',
      middle: 'T'
    }
  });
  console.log('what is entry', entry);
  const newId = new mongoose.Types.ObjectId();
  const newItem = { _id: newId, first: 'John', last: 'Johnson', middle: 'J' }
  const update = await Test.updateOne({
    _id: entry._id, 'items._id': id
  }, { $set: { 'items.$': { newItem } } });
  console.log('what is update', update);

  const res = await Test.findOne();
  console.log('what is res', res);
  console.log('what is newItem', res.items[0].newItem, typeof res.items[0].newItem, Array.isArray(res.items[0].newItem));

}

run();
```

**Comment by donoftime2018** [comment]
> Please modify the script below to demonstrate your issue
> 
> ```js
> const mongoose = require('mongoose');
> 
> const testSchema = new mongoose.Schema({
>   items: [{
> 
>   }]
> });
> const id = new mongoose.Types.ObjectId();
> const Test = mongoose.model('Test', testSchema);
> 
> async function run() {
>   await mongoose.connect('mongodb://localhost:27017');
>   await mongoose.connection.dropDatabase();
> 
>   const entry = await Test.create({
>     items: {
>       _id: id,
>       first: 'Test',
>       last: 'Testeron',
>       middle: 'T'
>     }
>   });
>   console.log('what is entry', entry);
>   const newId = new mongoose.Types.ObjectId();
>   const newItem = { _id: newId, first: 'John', last: 'Johnson', middle: 'J' }
>   const update = await Test.updateOne({
>     _id: entry._id, 'items._id': id
>   }, { $set: { 'items.$': { newItem } } });
>   console.log('what is update', update);
> 
>   const res = await Test.findOne();
>   console.log('what is res', res);
>   console.log('what is newItem', res.items[0].newItem, typeof res.items[0].newItem, Array.isArray(res.items[0].newItem));
> 
> }
> 
> run();
> ```In the $set, you should probably access the specific index being updated (i.e. { $set: { items[index]: { newItem } } }).  I think you might accidentally be updating the array instead of one specific object. Try seeing if you are accessing the items array in Test schema (i.e. you aren't creating multiple Test with different arrays). The best bet would be to check how you are accessing one array, and how you access the item being updated within it.

**Comment by kimbylr** [comment]
Thanks for your replies. I managed to pin it down:

### Script to reproduce

```
const itemSchema = new mongoose.Schema({ first: String, last: String });
const testSchema = new mongoose.Schema({ items: [itemSchema] });
const Test = mongoose.model('Test', testSchema);

export const testDB = async () => {
  await initDB();

  const entryId = new mongoose.Types.ObjectId();
  const firstItemId = new mongoose.Types.ObjectId();
  const entry = await Test.create({
    _id: entryId,
    items: [{ _id: firstItemId, last: 'Test', first: 'T' }],
  });
  console.log('entry', entry);

  // add 2nd item
  const item2 = { last: '222', first: '222' };
  await Test.updateOne({ _id: entryId }, { $push: { items: item2 } });

  //edit
  const newItem = { _id: firstItemId, last: 'John', first: 'J' };
  console.log('newItem', newItem);
  const update = await Test.updateOne(
    { _id: entryId, 'items._id': firstItemId },
    { $set: { 'items.$': newItem } },
  );
  console.log('update', update);

  const res = await Test.findOne({ _id: entryId });
  console.log('what is res', res);
  console.log('what is newItem', res?.items[0], typeof res?.items[0], Array.isArray(res?.items[0]));
};
```

### console output

```
entry {
  items: [
    {
      first: 'T',
      last: 'Test',
      _id: new ObjectId("6539911d9ad948a41698d442")
    }
  ],
  _id: new ObjectId("6539911d9ad948a41698d441"),
  __v: 0
}

newItem {
  _id: new ObjectId("6539911d9ad948a41698d442"),
  last: 'John',
  first: 'J'
}

update {
  acknowledged: true,
  modifiedCount: 1,
  upsertedId: null,
  upsertedCount: 0,
  matchedCount: 1
}

what is res {
  _id: new ObjectId("6539911d9ad948a41698d441"),
  items: [
    { '0': [Object], _id: new ObjectId("6539911d9ad948a41698d44b") },
    {
      first: '222',
      last: '222',
      _id: new ObjectId("6539911d9ad948a41698d446")
    }
  ],
  __v: 0
}

what is newItem {
  '0': {
    first: 'J',
    last: 'John',
    _id: new ObjectId("6539911d9ad948a41698d442")
  },
  _id: new ObjectId("6539911d9ad948a41698d44b")
} object false
```

### notes

The array seems to be casted to an object, which is why `Array.isArray`returns false. But looking at the document directly it's quite clear it's an array:
![image](https://github.com/Automattic/mongoose/assets/21242860/e970ca64-e651-411c-9b0f-539538352739)

This is the outcome in 7.6.2. In 7.6.1, it doesn't happen. It also doesn't happen if there's only one object in the array.

It might very well be that I'm doing something weird (not that familiar with `$put` etc. commands), or that the new behaviour is in fact the expected one. In any case, it's different.

**Comment by IslandRhythms** [comment]
You must look in the db to see the bug. The logs are misleading.

```javascript
const mongoose = require('mongoose');

const itemSchema = new mongoose.Schema({ first: String, last: String });
const testSchema = new mongoose.Schema({ items: [itemSchema] });
const Test = mongoose.model('Test', testSchema);


async function run() {
  await mongoose.connect('mongodb://localhost:27017/test14000');
  await mongoose.connection.dropDatabase();

  const entryId = new mongoose.Types.ObjectId();
  const firstItemId = new mongoose.Types.ObjectId();
  const entry = await Test.create({
    _id: entryId,
    items: [{ _id: firstItemId, last: 'Test', first: 'T' }],
  });
  console.log('entry', entry);

  // add 2nd item
  const item2 = { last: '222', first: '222' };
  await Test.updateOne({ _id: entryId }, { $push: { items: item2 } });

  //edit
  const newItem = { _id: firstItemId, last: 'John', first: 'J' };
  console.log('newItem', newItem);
  const update = await Test.updateOne(
    { _id: entryId, 'items._id': firstItemId },
    { $set: { 'items.$': newItem } },
  );
  console.log('update', update);

  const res = await Test.findOne({ _id: entryId });
  console.log('what is res', res);
  console.log('what is newItem', res?.items[0], typeof res?.items[0], Array.isArray(res?.items[0]));
}

run();
```

**Comment by donoftime2018** [comment]
It might be with how you are accessing the object in the array. Maybe do an
Object typecast in the statement where the object in the list is updated?
Maybe check the logic for how you are accessing that particular object?

**Comment by donoftime2018** [comment]
How can I access the repository?

**Comment by donoftime2018** [comment]
OK. I got it working. All I did was replace $set with $setOnInsert.

**Comment by donoftime2018** [comment]
const mongoose = require('mongoose');

const itemSchema = new mongoose.Schema({ first: String, last: String });
const testSchema = new mongoose.Schema({ items: [itemSchema] });
const Test = mongoose.model('Test', testSchema);

async function run() {
    await mongoose.connect('mongodb://localhost:27017/test14000');
    await mongoose.connection.dropDatabase();
  
    const entryId = new mongoose.Types.ObjectId();
    const firstItemId = new mongoose.Types.ObjectId();
    const entry = await Test.create({
      _id: entryId,
      items: [{ _id: firstItemId, last: 'Test', first: 'T' }],
    });
    console.log('entry', entry);
  
    // add 2nd item
    const secondItemId = new mongoose.Types.ObjectId();
    const item2 = { last: '222', first: '222', _id: secondItemId};

   const addToList = await Test.updateOne({ _id: entryId }, { $push: { items: item2 } });
   console.log(addToList, typeof(addToList))

   const list = await Test.findOne({_id: entryId})

   console.log(list)
  
    //edit
    const newItem = { _id: firstItemId, last: 'John', first: 'J' };
    console.log('newItem', newItem);
    const update = await Test.updateOne(
      { _id: entryId, 'items._id': firstItemId },
      { $setOnInsert: { 'items.$': newItem } },
    );
    console.log('update', update, typeof(update));
  
    const res = await Test.findOne({ _id: entryId });
    console.log('what is res', res);
    console.log('what is newItem', res?.items[0], typeof res?.items[0]);
    console.log('what is newItem', res?.items[1], typeof res?.items[1]);

    console.log(res?.items)
  }
  run()

**Comment by vkarpov15** [comment]
I took a look and confirmed that this was a bug in 7.6.2-7.6.4, but this is fixed in 7.6.5 with #14066. Please upgrade to 7.6.5 for the fix :+1:
